x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java

```diff
@@ -0,0 +1,158 @@
+    private final Client client;
+    private volatile int maxFailureRetries;
+
+    public ResultsPersisterService(Client client, ClusterService clusterService, Settings settings) {
+        this.client = client;
+        this.maxFailureRetries = PERSIST_RESULTS_MAX_RETRIES.get(settings);
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(PERSIST_RESULTS_MAX_RETRIES, this::setMaxFailureRetries);
+    }
+
+    void setMaxFailureRetries(int value) {
+        this.maxFailureRetries = value;
+    }
+
+    public BulkResponse indexWithRetry(String jobId,
+                                       String indexName,
+                                       ToXContent object,
+                                       ToXContent.Params params,
+                                       WriteRequest.RefreshPolicy refreshPolicy,
+                                       String id,
+                                       Supplier<Boolean> shouldRetry) throws IOException {
+        BulkRequest bulkRequest = new BulkRequest().setRefreshPolicy(refreshPolicy);
+        try (XContentBuilder content = object.toXContent(XContentFactory.jsonBuilder(), params)) {
+            bulkRequest.add(new IndexRequest(indexName).id(id).source(content));
+        }
+        return bulkIndexWithRetry(bulkRequest, jobId, shouldRetry);
+    }
+
+    public BulkResponse bulkIndexWithRetry(BulkRequest bulkRequest, String jobId, Supplier<Boolean> shouldRetry) {
+        int currentMin = MIN_RETRY_SLEEP_MILLIS;
+        int currentMax = MIN_RETRY_SLEEP_MILLIS;
+        int currentAttempt = 0;
+        BulkResponse bulkResponse = null;
+        while(currentAttempt <= maxFailureRetries) {
+            bulkResponse = bulkIndex(bulkRequest);
+            if (bulkResponse.hasFailures() == false) {
+                return bulkResponse;
+            }
+            if (shouldRetry.get() == false) {
+                throw new ElasticsearchException("[{}] failed to index all results. {}", jobId, bulkResponse.buildFailureMessage());
+            }
+            if (currentAttempt > maxFailureRetries) {
+                LOGGER.warn("[{}] failed to index after [{}] attempts. Setting [xpack.ml.persist_results_max_retries] was reduced",
+                    jobId,
+                    currentAttempt);
+                throw new ElasticsearchException("[{}] failed to index all results after [{}] attempts. {}",
+                    jobId,
+                    currentAttempt,
+                    bulkResponse.buildFailureMessage());
+            }
+            currentAttempt++;
+            // Since we exponentially increase, we don't want force randomness to have an excessively long sleep
+            if (currentMax < MAX_RETRY_SLEEP_MILLIS) {
+                currentMin = currentMax;
+            }
+            double backOff = ((1 << Math.min(currentAttempt, MAX_RETRY_EXPONENT)) - 1) / 2.0;
+            int max = (int)(backOff * 100);
```

Comment:
There is undocumented subtlety here.  `backOff * 100` can be greater than `Integer.MAX_VALUE`, and then the cast to `int` of a `double` greater than `Integer.MAX_VALUE` will result in `Integer.MAX_VALUE`.

But it makes me wonder whether it would be clearer to just use:

```
    int uncappedBackOff = ((1 << Math.min(currentAttempt, MAX_RETRY_EXPONENT)) - 1) * (100 / 2);
```

and change `MAX_RETRY_EXPONENT` to 24.

This avoids any subtlety with casting `int` to `double` and back again.  Or if there is a really good reason to go via `double`, please comment it.

Reply:
Rounding up and `* 50` should work. Let me experiment